[gh issue/pr comment] Add support for last comment delete for issues and PRs

sinansonmez opened the pull request:
Fixes #10366

Add `--delete-last` option to issue/pr comment.
This option deletes the last comment that you are the author of

sinansonmez committed:
deletion for issues with confirmation flag

sinansonmez committed:
add handling for interaction case

sinansonmez committed:
finish implementation for issues

sinansonmez committed:
finish the implementation for issues

sinansonmez committed:
finalize the implementation for PR

sinansonmez committed:
fix missing --yes flag for PR

sinansonmez commented on api/queries_comments.go:
```diff
@@ -99,6 +103,27 @@ func CommentUpdate(client *Client, repoHost string, params CommentUpdateInput) (
 
+func CommentDelete(client *Client, repoHost string, params CommentDeleteInput) (error) {
+	var mutation struct {
+		DeleteIssueComment struct {
```
1 clarification. The mutation mention `issue` in the name. However, it works both for issues and PRs. Is it OK to use for both? Or for PRs should I use something else?

babakks commented on pkg/cmd/pr/comment/comment.go:
```diff
@@ -76,6 +77,8 @@ func NewCmdComment(f *cmdutil.Factory, runF func(*shared.CommentableOptions) err
 	cmd.Flags().BoolP("web", "w", false, "Open the web browser to write the comment")
 	cmd.Flags().BoolVar(&opts.EditLast, "edit-last", false, "Edit the last comment of the same author")
+	cmd.Flags().BoolVar(&opts.DeleteLast, "delete-last", false, "Delete the last comment of the same author")
+	cmd.Flags().BoolVar(&opts.Confirmed, "yes", false, "Skip the confirmation prompt")
```
The doc should be clear that this option corresponds to `--delete-last`. Something like "Skip the delete confirmation prompt when `--delete-last` is provided".

babakks commented on pkg/cmd/issue/comment/comment.go:
```diff
@@ -70,6 +71,8 @@ func NewCmdComment(f *cmdutil.Factory, runF func(*prShared.CommentableOptions) e
 	cmd.Flags().BoolP("web", "w", false, "Open the web browser to write the comment")
 	cmd.Flags().BoolVar(&opts.EditLast, "edit-last", false, "Edit the last comment of the same author")
+	cmd.Flags().BoolVar(&opts.DeleteLast, "delete-last", false, "Delete the last comment of the same author")
+	cmd.Flags().BoolVar(&opts.Confirmed, "yes", false, "Skip the confirmation prompt")
```
The same.

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -41,11 +42,14 @@ type CommentableOptions struct {
 	Body                      string
 	EditLast                  bool
+	DeleteLast                bool
+	Confirmed                 bool
```
Let's make this field name more explicit. How about `DeleteLastConfirmed`?

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -41,11 +42,14 @@ type CommentableOptions struct {
 	InteractiveEditSurvey     func(string) (string, error)
 	ConfirmSubmitSurvey       func() (bool, error)
 	ConfirmCreateIfNoneSurvey func() (bool, error)
+	ConfirmDeleteComment      func(string) (bool, error)
```
This can be more explicit like `ConfirmDeleteLastComment`.

babakks commented on pkg/cmd/pr/comment/comment.go:
```diff
@@ -76,6 +77,8 @@ func NewCmdComment(f *cmdutil.Factory, runF func(*shared.CommentableOptions) err
 	cmd.Flags().BoolP("editor", "e", false, "Skip prompts and open the text editor to write the body in")
 	cmd.Flags().BoolP("web", "w", false, "Open the web browser to write the comment")
 	cmd.Flags().BoolVar(&opts.EditLast, "edit-last", false, "Edit the last comment of the same author")
+	cmd.Flags().BoolVar(&opts.DeleteLast, "delete-last", false, "Delete the last comment of the same author")
```
The "the same author" is a bit confusing. Probably it's better to point to "the current user".

sinansonmez replied:
Thanks @babakks  , I wanted to align the wording with "EditLast" warning. If you agree, I changed both to "the current user" to be consistent

babakks replied:
Nice.

babakks commented on pkg/cmd/issue/comment/comment.go:
```diff
@@ -70,6 +71,8 @@ func NewCmdComment(f *cmdutil.Factory, runF func(*prShared.CommentableOptions) e
 	cmd.Flags().BoolP("editor", "e", false, "Skip prompts and open the text editor to write the body in")
 	cmd.Flags().BoolP("web", "w", false, "Open the web browser to write the comment")
 	cmd.Flags().BoolVar(&opts.EditLast, "edit-last", false, "Edit the last comment of the same author")
+	cmd.Flags().BoolVar(&opts.DeleteLast, "delete-last", false, "Delete the last comment of the same author")
```
The same.

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -225,6 +232,50 @@ func updateComment(commentable Commentable, opts *CommentableOptions) error {
+				return errDeleteNotConfirmed
+			}
+		} else {
+			return errors.New("you must use --yes flag to confirm deletion in non-interactive mode")
```
Flag/option validation should be done in `CommentablePreRun`. This way we can confirm the validation is working as expected in `TestNewCmdComment` functions (in `pkg/cmd/pr/comment/comment_test.go` and `pkg/cmd/issue/comment/comment_test.go`).

I think these are the checks we need to perform:
- `--delete-last` and `--yes` must be provided in non-interactive mode. To avoid confusion around `opts.Interactive`, here non-interactive means a non-TTY IO (i.e., `opts.IO.CanPrompt() == true`). 
- `--delete-last` cannot be used with `--body`, `--body-file`, `--editor`, `--web`, `--edit-last` or `--create-if-none` flags.

Also, since `CommentablePreRun` decision making will now grow more complicated, we might want to add a test to make sure it behaves as we expect. I can help with that part when we get to it.

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -225,6 +232,50 @@ func updateComment(commentable Commentable, opts *CommentableOptions) error {
+		return errNoUserComments
+	}
+
+	lastComment := &comments[len(comments)-1]
```
No need to grab a pointer:

```suggestion
	lastComment := comments[len(comments)-1]
```

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -225,6 +232,50 @@ func updateComment(commentable Commentable, opts *CommentableOptions) error {
+			}
+			if !ok {
+				return errDeleteNotConfirmed
+			}
```
**Suggestion:** you can fold this into:
```suggestion
			if ok, err := opts.ConfirmDeleteComment(lastComment.Body); err != nil {
				return err
			} else if !ok {
				return errDeleteNotConfirmed
			}
```

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -225,6 +232,50 @@ func updateComment(commentable Commentable, opts *CommentableOptions) error {
+	cs := opts.IO.ColorScheme()
+
+	if (opts.IO.CanPrompt() && !opts.Confirmed) {
+		if opts.Interactive {
```
As of the current implementation of `CommentablePreRun`, `opts.Interactive` indicates whether a comment text is not provided (i.e., none of `--body`, `--body-file`, `--editor` or `--web` is passed).

In case of the `deleteComment` function, no comment text is expected, so we don't need to check `opts.Interactive`.

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -225,6 +232,50 @@ func updateComment(commentable Commentable, opts *CommentableOptions) error {
+	}
+
+	if !opts.Quiet {
+		fmt.Fprintln(opts.IO.Out, "Deleted the comment successfully.")
```
Two things:
- This should be printed on `opts.IO.ErrOut`.
- The message reads better as "Comment deleted"

```suggestion
		fmt.Fprintln(opts.IO.ErrOut, "Comment deleted")
```

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -260,6 +311,12 @@ func CommentableEditSurvey(cf func() (gh.Config, error), io *iostreams.IOStreams
 
+func CommentableConfirmDeleteComment(p Prompt) func(string) (bool, error) {
+	return func(body string) (bool, error) {
+		return p.Confirm(fmt.Sprintf("Deleting the comment with a body: '%s'.", body), true)
```
I think the prompt format reads better as `"Delete the comment: %s?"`

However, multiline/long comments might create a unpleasant experience here. I'll put a comment down here to discuss it separately.

babakks reviewed (commented):
Great PR! Thanks @sinansonmez! It looks very good, just needs some changes.

I also concerned about how to display the comment for confirmation, especially for long/multiline comments, but that's another comment that needs some discussion.

babakks commented:
While reviewing this, I was wondering "what should we do when confirming the operation?". Probably one of these is the answer:

1. Ask "Delete comment?".
   - This doesn't help much because the user might want to see the comment before proceeding.
1. Ask "Delete comment: "the entire comment text"? 
   - Long/multiline comments mess up with this one.
1. Ask "Delete comment: "truncated comment text..."? 
   - Multiline comments can still mess with this, unless we quote them (i.e., printf with "%q").
1. Print the comment text and then a short prompt like "Delete comment?"?
   - How about markdown rendering? do we need it?

@williammartin WDYT?

sinansonmez committed:
Merge branch 'trunk' into feature/10366/issue-comment-delete-last

sinansonmez committed:
address PR comments related to feedbacks

babakks commented on pkg/cmd/pr/comment/comment.go:
```diff
@@ -76,9 +76,9 @@ func NewCmdComment(f *cmdutil.Factory, runF func(*shared.CommentableOptions) err
-	cmd.Flags().BoolVar(&opts.Confirmed, "yes", false, "Skip the confirmation prompt")
+	cmd.Flags().BoolVar(&opts.EditLast, "edit-last", false, "Edit the last comment of the current user")
+	cmd.Flags().BoolVar(&opts.DeleteLast, "delete-last", false, "Delete the last comment of the current user")
+	cmd.Flags().BoolVar(&opts.DeleteLast, "yes", false, "Skip the delete confirmation prompt when --delete-last is provided")
```
This should be `opts.DeleteLastConfirmed`.

sinansonmez replied:
thanks for the catch

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -17,7 +17,8 @@ import (
 )
 
-var errNoUserComments = errors.New("no comments found for current user")
+var errNoUserComments = errors.New("no comments found for the current user")
```
Since this error was already there I'd just leave it as is, in case some user is already  `grep`-ing against the exact message.

sinansonmez replied:
thanks, reverted. I will keep in my for future changes

babakks reviewed (commented):
Thanks for the changes, @sinansonmez!

Could you also please run `gofmt` to make sure it's formatted as it should be?

Have you already tried the `--delete-last` and `--yes` options together on a real issue/PR? Because, as I'm trying, I still get the prompt:

```
$ go run ./cmd/gh pr comment -R babakks/dummy-gh-pr 1 --delete-last --yes
! Deleted comments cannot be recovered.
? Delete the comment: This is a test comment.? Yes
```

(Feel free to try it on my test repo above.)

Also, running it with a non-TTY stdout results in an error:

```
$ go run ./cmd/gh pr comment -R babakks/dummy-gh-pr 1 --delete-last --yes | cat
flags required when not running interactively
...
```

As I mentioned before, the `CommentablePreRun` function is responsible for flag/option validation and it needs to be updated with the new changes. This also means we need to revisit the tests, but let's do that as the final step.

sinansonmez committed:
improve CommentablePreRun for pre checks

sinansonmez committed:
Merge branch 'trunk' into feature/10366/issue-comment-delete-last

sinansonmez commented:
@babakks thanks for the 2nd round of review. 

When you reviewed it, my changes were not final. I hope it is now ready for your review. I also tested it by using one of my test repo, it seems like it is working as expected now.

As suggested
- I improved `CommentablePreRun` to check non-TTY and `DeleteLastConfirmed`
- I also improved `TestNewCmdComment` by adding `isTTY` to have a finer control over the environment (i.e. TTY or no-TTY) that we are running tests

Let me know if there is any place which could be improved

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -78,6 +78,13 @@ func CommentablePreRun(cmd *cobra.Command, opts *CommentableOptions) error {
+		if opts.IO.CanPrompt() || opts.DeleteLastConfirmed {
+			return nil
+		}
+		return cmdutil.FlagErrorf("you must use `--yes` flag to confirm deletion in non-interactive mode")
```
Maybe rephrase the message like this: "should provide `--yes` to confirm deletion in non-interactive mode".

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -78,6 +78,13 @@ func CommentablePreRun(cmd *cobra.Command, opts *CommentableOptions) error {
+	if opts.DeleteLast {
+		if opts.IO.CanPrompt() || opts.DeleteLastConfirmed {
+			return nil
+		}
```
We also need to make sure no input is provided in the case of `opts.DeleteLast` being `true`:

```suggestion
	if opts.DeleteLast {
		if inputFlags > 0 {
			return cmdutil.FlagErrorf("should not provide comment body when using `--delete-last`")
		}
		if opts.IO.CanPrompt() || opts.DeleteLastConfirmed {
			return nil
		}
```

sinansonmez replied:
thanks for the good catch. I also wrote the test for this case

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -271,6 +325,12 @@ func CommentableEditSurvey(cf func() (gh.Config, error), io *iostreams.IOStreams
 
+func CommentableConfirmDeleteLastComment(p Prompt) func(string) (bool, error) {
+	return func(body string) (bool, error) {
+		return p.Confirm(fmt.Sprintf("Delete the comment: %s?", body), true)
```
Let's quote the body with `%q`.

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -236,6 +250,46 @@ func updateComment(commentable Commentable, opts *CommentableOptions) error {
+
+	if opts.IO.CanPrompt() && !opts.DeleteLastConfirmed {
+		fmt.Fprintf(opts.IO.Out, "%s Deleted comments cannot be recovered.\n", cs.WarningIcon())
+		ok, err := opts.ConfirmDeleteLastComment(lastComment.Body)
```
As an initial solution, let's strip the comment body if it's more than 40 chars long (and add "..." at the end if so). This is to avoid printing lots of text if the comment is a big one, which is not a very accessible experience.

But we need to cover this path in the tests. There are already tests that cover the interactive confirmation path, but we need to have two test cases; shorter and longer than 40 chars. For these test cases we need to assert the prompt message against the expected value.

Now that I'm looking at the `Test_commentRun` in `pkg/cmd/pr/comment/comment_test.go`, we might want to introduce an override to the table test type and populate it only in these new test cases. Something like:

```go
tests := []struct {
    name          string
	input         *shared.CommentableOptions
	emptyComments bool
    comments      []api.Comments                          // << this one
	httpStubs     func(*testing.T, *httpmock.Registry)
	expectErr     bool
	stdout        string
	stderr        string
	wantsErr      bool
}
```

And down in the loop we pick it up like this:

```go
comments := api.Comments{Nodes: []api.Comment{
    {ID: "id1", Author: api.CommentAuthor{Login: "octocat"}, URL: "https://github.com/OWNER/REPO/pull/123#issuecomment-111", ViewerDidAuthor: true},
    {ID: "id2", Author: api.CommentAuthor{Login: "monalisa"}, URL: "https://github.com/OWNER/REPO/pull/123#issuecomment-222"},
}}
if tt.emptyComments {
    comments.Nodes = []api.Comment{}
} else if tt.comments != nil {
    comments = tt.comments
}
```

This way we can replace arbitrary short/long comment bodies for those new test cases.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -451,6 +513,77 @@ func Test_commentRun(t *testing.T) {
+				DeleteLast:          true,
+				DeleteLastConfirmed: true,
+
+				ConfirmDeleteLastComment: func(string) (bool, error) { return true, nil },
```
Here, we should assert the given argument (i.e., body) is the expected value. I mean:

```go
ConfirmDeleteLastComment: func(body string) (bool, error) {
    if body == "expected" {
        return false, errors.New("unexpected comment body")
    }
    return true, nil
},
```

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -451,6 +513,77 @@ func Test_commentRun(t *testing.T) {
+				Interactive: true,
+				DeleteLast:  true,
+
+				ConfirmDeleteLastComment: func(string) (bool, error) { return true, nil },
```
The same.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -451,6 +513,77 @@ func Test_commentRun(t *testing.T) {
+				Interactive: true,
+				DeleteLast:  true,
+
+				ConfirmDeleteLastComment: func(string) (bool, error) { return false, nil },
```
The same.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -489,6 +622,7 @@ func Test_commentRun(t *testing.T) {
 			err := shared.CommentableRun(tt.input)
 			if tt.wantsErr {
 				assert.Error(t, err)
+				assert.Equal(t, tt.stderr, err.Error())
```
This should be:

```go
assert.Equal(t, tt.stderr, stderr)
```

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -150,52 +162,98 @@ func TestNewCmdComment(t *testing.T) {
+			isTTY := tt.isTTY
+			ios.SetStdoutTTY(isTTY)
+			ios.SetStdinTTY(isTTY)
+			ios.SetStderrTTY(isTTY)
```
(Doing it here since I cannot comment on further lines)

We should update the assertions at the end of this loop to cover the `DeleteLast` and `DeleteConfirmed` fields, otherwise these test cases will never fail.

sinansonmez replied:
thanks for the catch

sinansonmez replied:
thanks for the catch

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -241,6 +299,7 @@ func Test_commentRun(t *testing.T) {
 		input         *shared.CommentableOptions
 		emptyComments bool
 		httpStubs     func(*testing.T, *httpmock.Registry)
+		expectErr     bool
```
Probably a left over, abandoned in favour of `wantsErr`.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -451,6 +513,77 @@ func Test_commentRun(t *testing.T) {
 			stdout: "https://github.com/OWNER/REPO/pull/123#issuecomment-456\n",
 		},
+		{
+			name: "non interactive editor with delete last without any comment",
```
To keep it consistent with other cases: "deleting last comment with non-interactive and with no comments"

Please apply the same for the other cases you've added.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -451,6 +513,77 @@ func Test_commentRun(t *testing.T) {
+			stderr: "Comment deleted\n",
+		},
+		{
+			name: "interactive editor with delete last and not confirmed default but confirmed interactively",
```
Should be "...but confirmation declined".

sinansonmez replied:
thx for the catch

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -150,52 +162,98 @@ func TestNewCmdComment(t *testing.T) {
+			wantsErr: false,
+		},
+		{
+			name:  "non-interactive delete last no flag",
```
To make it clearer: "non-interactive delete last without confirmation". Please apply to the rest of the newly added cases.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -150,52 +162,98 @@ func TestNewCmdComment(t *testing.T) {
+			wantsErr: false,
+		},
+		{
+			name:  "interactive delete last flag",
```
We also need a couple of more test cases to cover the invalid cases where `--delete-last` is present with one of the comment body flags (i.e., `--body`, `--body-file`, `--editor`, or `--web`).

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -74,6 +78,13 @@ func CommentablePreRun(cmd *cobra.Command, opts *CommentableOptions) error {
 		return cmdutil.FlagErrorf("`--create-if-none` can only be used with `--edit-last`")
 	}
 
+	if opts.DeleteLast {
```
We should also cover the case that `--yes` is provided without `--delete-last`. It's possible that a user thinks they can provide `--yes` to avoid interaction for other commands. A standalone `if`-statement like this will do the trick:

```go
if opts.DeleteLastConfirmed && !opts.DeleteLast {
    return cmdutil.FlagErrorf("`--yes` should only be used with `--delete-last`")
}
```

We also need a test case to confirm this behaviour as well.

sinansonmez replied:
thanks again for another edge case catch

babakks reviewed (commented):
Thanks for the changes! :pray: And sorry for the premature review. My bad. 

Now we need to polish the tests and make sure the behaviour is stable enough. So close to land this. :tada:

I went over the `pr comment` tests and put a bunch of comments. The `issue comment` tests need the same changes, so I didn't duplicate my comments.

Note that I've also asked for truncating long comments, because I think it's important to have UX and accessibility in our sight. Anyway, let me know if I can help with impl or writing tests for it (or any other part of the changes). I'd be happy to help.

sinansonmez committed:
improve confirmation prompt and truncate long comment body

sinansonmez committed:
address PR comments on tests

sinansonmez commented:
@babakks Thanks for such a deep review. hopefully, I addressed all of your points.  Please let me know if I missed something or addressed something incorrectly

babakks commented on pkg/cmd/issue/comment/comment_test.go:
```diff
@@ -494,27 +537,27 @@ func Test_commentRun(t *testing.T) {
 		},
 		{
-			name: "non interactive editor with delete last without any comment",
+			name: "deleting last comment with non interactive editor with delete last without any comment",
```
"editor" is irrelevant for these tests.

babakks commented on pkg/cmd/pr/comment/comment_test.go:
```diff
@@ -514,27 +557,27 @@ func Test_commentRun(t *testing.T) {
 		},
 		{
-			name: "non interactive editor with delete last without any comment",
+			name: "deleting last comment with non interactive editor with delete last without any comment",
```
And these.

babakks commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -262,7 +269,7 @@ func deleteComment(commentable Commentable, opts *CommentableOptions) error {
 	if opts.IO.CanPrompt() && !opts.DeleteLastConfirmed {
 		fmt.Fprintf(opts.IO.Out, "%s Deleted comments cannot be recovered.\n", cs.WarningIcon())
-		ok, err := opts.ConfirmDeleteLastComment(lastComment.Body)
+		ok, err := opts.ConfirmDeleteLastComment(text.Truncate(40, lastComment.Body))
```
This doesn't work for our use case because it just truncates long lines and not the string as a whole. The best is to do an ordinary slice range indexing. I know it's not the best way to do it, but I think it's okay.

babakks reviewed (commented):
Thanks for the changes, @sinansonmez! and also for picking up this issue. :pray: 

I have a couple of comments which I'll just now apply myself and go for merging this PR.

babakks committed:
Truncate comment for confirmation prompt

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
Improve test case descriptions

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
Fix mock comment body

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
Remove irrelevant prompt stub

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
Use `opts.Interactive` as TTY indicator

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks commented:
I did a bunch of small/cosmetic changes and I'm now waiting for the CI to finish its job.

sinansonmez commented:
@babakks  Thanks for doing the last mile changes directly and your support during review. 

btw I am willing to take more issues. I checked help-wanted issues but there isn’t any available one currently. Feel free to assign if something comes up

babakks committed:
Fix expected `Interactive` value

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks committed:
Polish `TestNewCmdComment`

Signed-off-by: Babak K. Shandiz <babakks@github.com>

babakks commented:
@sinansonmez No worries. It's my pleasure. :heart: 

We're currently reviewing our internal process about `help-wanted` issues and decided to temporarily put a hold on them. Hopefully, you'll soon see new `help-wanted` issues popping up.
